dynamiq/nodes/agents/orchestrators/graph.py

```diff
@@ -0,0 +1,531 @@
+                    input_data={"input": agent_input},
+                    config=config,
+                    run_depends=run_depends,
+                    **kwargs,
+                )
+
+                result = response.output.get("content")
+
+                if response.status != RunnableStatus.SUCCESS:
+                    logger.error(f"GraphOrchestrator: Failed to execute Agent {task.name} with Error: {result}")
+                    raise OrchestratorError(f"Failed to execute Agent {task.name} with Error: {result}")
+
+                return result, {}
+
+            except Exception as e:
+                logger.error(f"GraphOrchestrator: Error when parsing response about next state {e}")
+                raise OrchestratorError(f"Error when parsing response about next state {e}")
+
+        elif isinstance(task, FunctionTool):
+            input_data = {"context": global_context | {"history": chat_history}}
+        else:
+            input_data = {**global_context, "history": chat_history}
+
+        response = task.run(input_data=input_data, config=config, run_depends=run_depends, **kwargs)
+
+        context = response.output.get("content")
+
+        if response.status != RunnableStatus.SUCCESS:
+            logger.error(f"GraphOrchestrator: Failed to execute {task.name} with Error: {context}")
+            raise OrchestratorError(f"Failed to execute {task.name} with Error: {context}")
+
+        if not isinstance(context, dict):
+            raise OrchestratorError(
+                f"Error: Task returned invalid data format. Expected a dictionary got {type(context)}"
+            )
+
+        if "result" not in context:
+            raise OrchestratorError("Error: Task returned dictionary with no 'result' key in it.")
+
+        if "history" in context:
+            context.pop("history")
+
+        result = context.pop("result")
+
+        return result, context
+
+    def execute(self, input_data: StateInputSchema, config: RunnableConfig = None, **kwargs) -> dict:
+        logger.debug(f"State {self.id}: starting the flow with input_task:\n```{input_data}```")
+        kwargs.pop("run_depends", None)
+        self.run_on_node_execute_run(config.callbacks, **kwargs)
+        kwargs = kwargs | {"parent_run_id": kwargs.get("run_id")}
+
+        global_context = input_data.context
+        chat_history = input_data.chat_history
+
+        if len(self.tasks) == 1:
+            result, context = self._submit_task(
+                self.tasks[0],
+                global_context,
+                chat_history,
+                config=config,
+                **kwargs,
+            )
+
+            chat_history.append(
+                {
+                    "role": "system",
+                    "content": f"Result: {result}",
+                }
+            )
+
+            global_context = global_context | context
+
+        elif len(self.tasks) > 1:
+
+            contexts = []
+
+            with ThreadPoolExecutor() as executor:
```

Comment:
Like this idea, but I see that currently, Agents don't support concurrent/parallel run of the same instance. 
We have a few instance attributes (self._intermediate_steps, self._run_depends, self._chat_history, etc) that operates with mutable objects and also reset during `execute` call. So few calls or `run`/`execute` break logic.

- Let's use consecutive calls for now.
- Please also revert tracing changes as well. We need to improve that not just for Agent purposes with minimal breaking changes